### ISSUE TITLE
Private API in Intersphinx inventory

### ISSUE THREAD

**Issue by mthuurne** [twisted/pydoctor#353]
Currently the produced `objects.inv` file contains information on all documentables, whether they are private or public. I'm not sure whether that's desired though. For example a tool like `doc2dash` might want to process only public API, see hynek/doc2dash#53.

At the moment, the full name of a documentable is sufficient to determine whether it is private or public. However, if we ever implement the ability to mark selected constructors as private (see #305), that information will not be available unless we explicitly export it.

The simplest solution would be to simply omit private documentables from the Intersphinx inventory. Would that cause problems for anyone? Perhaps there would need to be a way to explicitly declare a documentable public even if its name suggests that it's private, for example if an originally private API became de-facto public since there was no other way for external projects to access essential functionality otherwise.

If there are use cases for having an inventory containing private API, maybe we should create two inventories: one with only public API and another with either only private API or public and private combined. The question would then be which inventory would be called `objects.inv` and which would get a new name.

My personal preference would be having only public API in `objects.inv` and either not including private API at all or putting that in a second inventory. But I don't want to cause too much disruption for pydoctor's users, so if there are scenarios where that approach would break things, please speak up.

**Comment by mthuurne** [comment]
Twisted links to private API in its documentation at least once (`twisted.trial._synctest._Assertions.assertWarns`), so not having an inventory that includes private API is not an option. Generating two inventories is possible though.

**Comment by mthuurne** [comment]
Since the Intersphinx extension itself uses the doc site URL as the key in its cache, we can't load more than one inventory for the same site. This means that if we pick the two inventories approach, we must have one inventory with only public API and another inventory with both public and private API, such that whichever inventory is selected by the user, only a single inventory has to be loaded.

**Comment by mthuurne** [comment]
To summarize, the current proposal is to produce two inventories:
- `objects.inv`, which includes only public API
- `objects-all.inv` (name not final), which includes both public and private API

**Comment by adiroiban** [comment]
+1 for having only public API in objects.inv  by default

objects.inv primary usage is for linking to external API, and you should only link to public external API.

----------

we can have an optional flat to also include private API when generating the objects.inv

For Twisted documenation build and public process I can imagine this process
cts.
* Build pydoctor and create an opbjects.inv with private API.
* Build Sphinx docs
* Re-run pydoctor to generate only the public API objects.inv
* Public the Sphinx docs / pydoctor API docs and objects.inv with only public API.

-----------

But for Twisted narratice docs, I would say that is better not to link to private API.
If links are needed like for `assertWarns , I think that `Assertions` should be exposed over a public module and link via that path.,

-----

So the summary:

* pydoctor should exclude private API by default
* there should be an option to include everything
* Twisted narrativedocs should only link to public API :)


Thanks!

**Comment by mthuurne** [comment]
I looked a bit more into Twisted's use of private API links and all of them point to methods of `twisted.trial._synctest._Assertions`. The one that was already mentioned is `compatibility-policy.rst` linking to `assertWarns()` and there are three links from `trial.rst` to `successResultOf()`, `failureResultOf()` and `assertNoResult()`. I agree that these methods should be considered public API, since `trial` is not feature complete without them.

We currently have a public method `assertWarns()` in a private class `_Assertions` in a private module `twisted.trial._synctest`. What mechanism should we use to declare them public?
- have pydoctor consider public methods from a private class that are inherited by a public class (`SynchronousTestCase`) to be public?
- make the class public in Twisted: rename from `_Assertions` to `Assertions`?
- make the module public in Twisted / move the class to a public module in Twisted?
- require both the class and the module to be public?

Maybe pydoctor could handle this situation as a re-export of API: by having a public class inherit public methods from a private class, those methods are effectively published by the public class. So `twisted.trial._synctest._Assertions.assertWarns()` would show up in the output as `twisted.trial.unittest.SynchronousTestCase.assertWarns()`.

Compared to how module members are re-exported, one thing I would want to do differently on the implementation side is to not actually move the methods, but instead insert an alias object that can be linked to into the model. That would allow re-exporting of methods in more than one public class.

---

However, in general a project could have documentation describing its implementation which would reference private API. So I do think that it is valid for a project to link to its own private API.

I don't really like the idea of having to run pydoctor twice to produce an `objects.inv` for internal use (public + private) and an `objects.inv` for publishing (public only). If we write two separate inventories instead, we'd only need one run of pydoctor and also only one output directory.

Or are you suggesting removing private API from the HTML as well unless a `--private` option is passed? That could make sense, but it's very different from how pydoctor works currently.

What is the purpose of private API documentation? Is it only for developers of the project itself or also for users under a "you can look, but don't touch" condition?

**Comment by adiroiban** [comment]
> have pydoctor consider public methods from a private class that are inherited by a public class (SynchronousTestCase) to be public? 

+1  - the method is on a public class, has a name of a public method.... it must be public.

And yes, this will resulting in re-exporting the public methods in more than one public class.


-----------

> Or are you suggesting removing private API from the HTML as well unless a --private option is passed? 

Good questions. No. I think that private API in HTML is ok.

So maybe, have it the other way. Have a `--only-public` flag that will generate the HTML and object inv only for public API.
If the docs make reference to private API, there will be no link created.

I feel that having the  object.inv without private API is a corner case and not something that the majority of pydoctor users will want.

------------

> What is the purpose of private API documentation? Is it only for developers of the project itself or also for users under a "you can look, but don't touch" condition?

For me, the private API documentation is just there to help get a better understanding  of  how things work...
but I can just read the source code.

I woulds say that that private  API documentation doesn't make much sense.

--------


My understanding is that we have this issue: objects.inv contains private API reference and this makes it hard to use it with 3rd  party tools as they break when names starting with underlines are found.

So we can focus on finding a fix for this.

One option is to just have  a separate generic object.inv filter that will take on inventory as input and product another one as output with all names filtered... this can be a simple tools outside of pydoctor

Cheers

**Comment by mthuurne** [comment]
> > have pydoctor consider public methods from a private class that are inherited by a public class (SynchronousTestCase) to be public?
> 
> +1 - the method is on a public class, has a name of a public method.... it must be public.
> 
> And yes, this will resulting in re-exporting the public methods in more than one public class.

Agreed. The more I think about it, the more this looks like the right solution.

> > Or are you suggesting removing private API from the HTML as well unless a --private option is passed?
> 
> Good questions. No. I think that private API in HTML is ok.
> 
> So maybe, have it the other way. Have a `--only-public` flag that will generate the HTML and object inv only for public API.
> If the docs make reference to private API, there will be no link created.
> 
> I feel that having the object.inv without private API is a corner case and not something that the majority of pydoctor users will want.

I think that the majority of users do not need private API in `objects.inv`, but at the same time they wouldn't notice if it is present. So what is the corner case depends on whether you look at whether the private API is required vs whether it is causing problems.

The use case where private API in an inventory is useful is when documenting your own project, in which case the name of the inventory doesn't matter, since Intersphinx can be given any file path. When linking to a published API, the name must be `objects.inv`. So that's a practical reason why, if we're going to produce two inventory files, `objects.inv` should be the public-only one.

> > What is the purpose of private API documentation? Is it only for developers of the project itself or also for users under a "you can look, but don't touch" condition?
> 
> For me, the private API documentation is just there to help get a better understanding of how things work...
> but I can just read the source code.
> 
> I woulds say that that private API documentation doesn't make much sense.

Meaning **rendered** private API documentation (HTML), right? Because the private documentation is useful when read together with the source code, similar to test case documentation.

> My understanding is that we have this issue: objects.inv contains private API reference and this makes it hard to use it with 3rd party tools as they break when names starting with underlines are found.

Not exactly: names with underscores are not a problem for the tools, the problem is that the private names cause a lot of noise in the output of the tool, making it harder for the user to find the right public API.

> So we can focus on finding a fix for this.
> 
> One option is to just have a separate generic object.inv filter that will take on inventory as input and product another one as output with all names filtered... this can be a simple tools outside of pydoctor

That's an interesting idea, but I'm not sure that Intersphinx as a format has all the information we need to decide which names are public. At some point, public/private may depend on more than a naming convention, for example it might be useful to mark a constructor as private if users are not supposed to manually instantiate a class.

**Comment by adiroiban** [comment]
> Meaning **rendered** private API documentation (HTML), right?

Yes. The rendered API documentation doesn't make sense.

-----------

> Not exactly: names with underscores are not a problem for the tools, the problem is that the private names cause a lot of noise in the output of the tool, making it harder for the user to find the right public API.

> That's an interesting idea, but I'm not sure that Intersphinx as a format has all the information we need to decide which names are public.

For now, filter only based on name starting with underscore... it's better that noting for now... to reduce the noise.
I don't think that intersphinx format has a flag for public/private

-------

So, in conclusion... things are complicated 

I don't know what is best :)

**Comment by mthuurne** [comment]
I just re-read #327 and noticed that we had come to the same conclusion there: that re-exporting the documentation for public methods implemented on a private class was the way to go.

**Comment by tristanlatr** [comment]
I think the latest option —privacy combined with option —make-intersphinx solves all the corners cases raised on this issue about intersphinx inventory generation. Meaning someone can rewrite privacy rules such that all stuff that was considered private is now considered hidden. So it won’t be part of the intersphinx.

I think this is a OK fix, I’m not comfortable dropping support for private stuff in the inventory by default. And generating two inventory every time seems unnecessary. 

Tell me what you think about the current workaround.

Let’s note that this does not solve the issue about methods inherited from private classes.
